rpcs3/Emu/Cell/Modules/cellMic.cpp

```diff
@@ -1,31 +1,398 @@
-#include "stdafx.h"
+﻿#include "stdafx.h"
 #include "Emu/System.h"
 #include "Emu/Cell/PPUModule.h"
+#include "Utilities/StrUtil.h"
 
 #include "cellMic.h"
 #include <Emu/IdManager.h>
 #include <Emu/Cell/lv2/sys_event.h>
 
 LOG_CHANNEL(cellMic);
 
-void mic_context::operator()()
+template <>
+void fmt_class_string<microphone_handler>::format(std::string& out, u64 arg)
 {
-	while (fxm::check<mic_thread>() == this && !Emu.IsStopped())
+	format_enum(out, arg, [](auto value)
 	{
-		thread_ctrl::wait_for(1000);
+		switch (value)
+		{
+		case microphone_handler::null: return "Null";
+		case microphone_handler::standard: return "Standard";
+		case microphone_handler::singstar: return "Singstar";
+		case microphone_handler::real_singstar: return "Real Singstar";
+		case microphone_handler::rocksmith: return "Rocksmith";
+		}
+
+		return unknown;
+	});
+}
 
+void mic_context::variable_byteswap(void* src, void* dst, u32 bytesize)
+{
+	switch (bytesize)
+	{
+	case 4:
+		*(u32*)dst = *(be_t<u32>*)src;
```

Comment:
You can also just say `dst = se_storage<u32>::swap(src)` btw. This is also fine; just dislike unnecessary indirection.

Reply:
I wasn't aware of that function at the time, changing it.